test/sslapitest.c

```diff
@@ -560,6 +561,103 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+static int send_recv_qery(SSL *clientssl, SSL *serverssl, int cfd, int sfd)
+{
+    static char count = 1;
+    char cbuf[16000] = {count++};
+    unsigned char sbuf[16000];
+    size_t err = 0;
+    char crec_seq_before[TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE];
+    char crec_seq_after[TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE];
+    char srec_seq_before[TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE];
+    char srec_seq_after[TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE];
+
+    memcpy(crec_seq_before, &clientssl->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    memcpy(srec_seq_before, &serverssl->rlayer.read_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    if (!TEST_true(SSL_write(clientssl, cbuf, sizeof(cbuf)) == sizeof(cbuf)))
+        goto end;
+
+    while ((err = SSL_read(serverssl, &sbuf, sizeof(sbuf))) != sizeof(sbuf)) {
+        if (SSL_get_error(serverssl, err) != SSL_ERROR_WANT_READ) {
+            goto end;
+        }
+    }
+
+    memcpy(crec_seq_after, &clientssl->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    memcpy(srec_seq_after, &serverssl->rlayer.read_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    /* verify the payload */
+    if (!TEST_true(memcmp(cbuf, sbuf, sizeof(cbuf)) == 0))
+        goto end;
+
+    /* kernel sequences were used instead of OpenSSL sequences */
+    if (!TEST_true(memcmp(crec_seq_before, crec_seq_after,
+                          TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE) == 0))
+        goto end;
+
+    if (!TEST_true(memcmp(srec_seq_before, srec_seq_after,
+                          TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE) != 0))
+        goto end;
+
+    return 1;
+end: 
+    return 0;
+}
+
+static int test_ktls(void)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    int cfd, sfd;
+
+    if (!TEST_true(create_test_sockets(&cfd, &sfd)))
+        goto end;
+
+    /* Create a session based on SHA-256 */
+    if (!TEST_true(create_ssl_ctx_pair(TLS_server_method(),
+                                       TLS_client_method(), &sctx,
+                                       &cctx, cert, privkey))
+            || !TEST_true(SSL_CTX_set_cipher_list(cctx,
+                                                  "ECDHE-RSA-AES128-GCM-SHA256"))
+            || !TEST_true(create_ssl_objects2(sctx, cctx, &serverssl,
+                                          &clientssl, sfd, cfd))
+            || !TEST_true(create_ssl_connection(serverssl, clientssl,
+                                                SSL_ERROR_NONE)))
+        goto end;
+
+    if (!TEST_true(BIO_get_ktls(clientssl->wbio))
+        || !TEST_true(BIO_get_ktls(serverssl->wbio)))
+        goto end;
+
+    if (!TEST_true(send_recv_qery(clientssl, serverssl, cfd, sfd))
+                || !TEST_true(send_recv_qery(clientssl, serverssl, cfd, sfd))
+                || !TEST_true(send_recv_qery(clientssl, serverssl, cfd, sfd)))
```

Comment:
Why do we do this 3 times?

Reply:
I wanted to verify that it doesn't break after running once. I'll remove the redundant two calls.